source/boundary_composition/function.cc

```diff
@@ -0,0 +1,219 @@
+/*
+  Copyright (C) 2011 - 2017 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/boundary_composition/function.h>
+#include <aspect/utilities.h>
+#include <aspect/global.h>
+#include <deal.II/base/signaling_nan.h>
+
+namespace aspect
+{
+  namespace BoundaryComposition
+  {
+    //template <int dim>
+    //Function<dim>::Function ()
+    //  :
+    //  boundary_composition_function (1)
+    //{}
+
+    template <int dim>
+    double
+    Function<dim>::
+    boundary_composition (const types::boundary_id /*boundary_indicator*/,
+                          const Point<dim> &position, const unsigned int n_comp) const
+    {
+      if (coordinate_system == ::aspect::Utilities::Coordinates::cartesian)
+        {
+          return function->value(position,n_comp);
+        }
+      else if (coordinate_system == ::aspect::Utilities::Coordinates::spherical)
+        {
+          const std_cxx11::array<double,dim> spherical_coordinates =
+            aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(position);
+          Point<dim> point;
+
+          for (unsigned int i = 0; i<dim; ++i)
+            point[i] = spherical_coordinates[i];
+
+          return function->value(point,n_comp);
+        }
+      else if (coordinate_system == Utilities::Coordinates::depth)
+        {
+          const double depth = this->get_geometry_model().depth(position);
+          Point<dim> point;
+          point(0) = depth;
+          return function->value(point,n_comp);
+        }
+      else
+        {
+          AssertThrow(false, ExcNotImplemented());
+          return numbers::signaling_nan<double>();
+        }
+    }
+
+
+    template <int dim>
+    void
+    Function<dim>::update()
+    {
+      // we get time passed as seconds (always) but may want
+      // to reinterpret it in years
+      if (this->convert_output_to_years())
+        boundary_composition_function.set_time (this->get_time() / year_in_seconds);
+      else
+        boundary_composition_function.set_time (this->get_time());
```

Comment:
you need to call this for the function object that you use in the `boundary_composition` function, not this one one (which is not needed).